### ISSUE TITLE
D3D12 profiling broken after 0.7.8?

### ISSUE THREAD

**Issue by ataulien-thqnordic** [wolfpld/tracy#271]
After successfully integrating D3D11 profiling, we went ahead and tried to integrate D3D12 profiling into our rendering path as well. Integrating mostly went smoothly, but at the end we were greeted with another Assertion in the TracyServer:

https://github.com/wolfpld/tracy/blob/b7a27d02afe1941eed8f64f2164447ff1a06daa6/server/TracyEvent.hpp#L209

We seem to get bogus values for the `end` parameter. After poking around in our codebase for some time, we noticed that the problem only occurs on the latest master version of tracy! After downgrading to the last official release 0.7.8, profiling D3D12 worked fine!

We confirmed that by integrating TracyD3D12 into a small test application. 

* Running on master? Assert. 
* Running on 0.7.8? Works fine.

Maybe there was an issue in the code that introduced the overflow-handling? (Commit https://github.com/wolfpld/tracy/commit/b7b9912b10c26bdac32490f4d12711ef8e111159)

**Comment by wolfpld** [comment]
Please specify your GPU and which drivers you use. Can you share the test application?

Can you verify that `if( tgpu < ctx->lastGpuTime )` branch is entered?

**Comment by ataulien-thqnordic** [comment]
I have forked the project that I've used for testing here: https://github.com/ataulien-thqnordic/Incarnation-Renderer/tree/tracy-gpu
The project used tracy for CPU profiling already, so I quickly added D3D12 profiling as well (`tracy-gpu` branch). If you want to try it, remember to also clone the submodules!

We had the issue on two different computers, both using NVidia 2070 SUPER on Windows 10, one with driver version 457.51 and one with 471.96.

I can confirm that the branch is being entered. If I just comment out the branch, it seems to run fine for a few frames, but then I get an assert here instead: https://github.com/wolfpld/tracy/blob/b7a27d02afe1941eed8f64f2164447ff1a06daa6/server/TracyWorker.cpp#L5586

**Comment by wolfpld** [comment]
Maybe this is related to stable power state setting?

> Note that due to artifacts from dynamic frequency scaling, GPU profiling may be slightly inaccurate. To counter this, `ID3D12Device::SetStablePowerState()` can be used to enable accurate profiling, at the expense of some performance. If the machine is not in developer mode, the device will be removed upon calling. Do not use this in shipping code.

See https://github.com/wolfpld/tracy/issues/227#issuecomment-851683471 for more information on this subject. The comments following one just linked show how broken GPU drivers can be and what issue the commit you originally referenced tries to fix.

Pinging @Xenonic, who did the implementation of D3D12 capture.

**Comment by Xenonic** [comment]
Thanks for letting me know, I'm seeing this issue on my end as well. I'll be able to take a look this weekend. @wolfpld have you tested OpenGL/Vulkan capture on master?

**Comment by wolfpld** [comment]
> could you explain a bit what the purpose of the overflow handling is for gpu timestamps?

See the issue I referenced above. It goes into great detail on why overflow handling was added.
tl;dr: drivers are shit, don't conform to spec, hacky workarounds needed.

> Is overflow handling meant to only occur when `m_refTimeGpu + ev.gpuTime` integer overflows?

No, the `tgpu` value has zero point at start of program execution and will have valid values for 1.6 days (2^47 ns). This variable does not overflow.

> If so, we can't just check `tgpu < ctx->lastGpuTime` since the scheduler could reorder work on the gpu such that zone times are out of the expected order.

That would certainly be a problem. Maybe there should be also a check for the clock going back in time for a reasonably small amount, in order to handle this properly?

**Comment by Xenonic** [comment]
Thanks for the info, this is quite an annoying situation. I tested the rolling back a small amount approach for ~10 minutes with no issues in d3d12, using this:
```
// If the timestamp is older than 2 seconds, assume the counter overflowed.
if( tgpu < ctx->lastGpuTime - 2000000000 )
```
Where the 2 seconds is the default timeout detection and recovery time on Nvidia. This should be less than any driver's overflow time period, while allowing for sufficiently slow async compute tasks plenty of time to complete.

**Comment by wolfpld** [comment]
It seems that the problem was resolved with the patch above. If it still persists, please reopen the issue and provide more information.
